### ISSUE TITLE
big array performance issue

### ISSUE THREAD

**Issue by masscrx** [angular-ui/ui-select#389]
Hi,
I have form for customer information, with three fields which use ui-select, every is for selecting country (country array length is 251) when countries array is loaded for ui-select it causes some lag on the page with this data, I've saw that there are some issues open with pagination for far future so now I can resolve my problem by adding checking for minimum input length for example 3 characters, than pagination is made on backend, but when customer is loaded with such information like $scope.customer.country_id I need to load all countries before, so solution with minium input length for 3 characters can't be applied in that case. Some ideas how to achieve that? I was trying to reduce watchers by adding :: to ui-select's expressions (one time binding - Angular 1.3 feauture) but with no result.

**Comment by dmccown1500** [comment]
this would be huge. I would not mind working on it at some point but that would require me to not be busy... so not any time soon I think

**Comment by maxisam** [comment]
+ 1.

**Comment by amcdnl** [comment]
What about virtual dom like react type implementation?

**Comment by dmccown1500** [comment]
Interesting that might do the trick. With a react style dom we could have
the list already built and merely injected when the ui-select dropdown
opens.

On a related note I did notice we have a lot of slowdowns that just come
from "ng-if/show"s that are calling a function. I think those have to call
every cycle right?

**Comment by Anton-Rodionov** [comment]
I also noticed ng-if slows ui-select down when it gets opened, but after quick profiling I found that it's because of *ngAnimate*, that by default "animates" every ng-if, ng-show, etc. (a lot of requestAnimationFrame calls). When I turned-off ngAnimate for ui-select it became 2x faster.

**Comment by dmccown1500** [comment]
If this could get merged in it would probably solve a lot of the issues as well by limiting the DOM size. https://github.com/angular-ui/ui-select/pull/430

**Comment by seiyria** [comment]
Any word on virtualizing? This is something that's come up for us as well.

**Comment by dmccown1500** [comment]
I have not done anything with it. infinite scrolling may still be the best bet and if not that then #430

**Comment by jpdesigndev** [comment]
@Anton-Rodionov How did you turn off ngAnimate on the ui-select. I, too, am getting rAF calls for every item in the list, which slows it down considerably.

**Comment by dorrotman** [comment]
Also interested in turning off ngAnimate effect on the ui-select component, while keeping it working for other components.

**Comment by yaring** [comment]
+1

**Comment by simison** [comment]
You guys might find some inspiration from here; https://github.com/select2/select2/issues/781 — since Select2 guys were having similar problems.

**Comment by wry791** [comment]
+1

**Comment by igorlino** [comment]
+1

**Comment by dmccown1500** [comment]
I have had recent success with infinite scroll using this library. http://binarymuse.github.io/ngInfiniteScroll/ it is a bit more work but if you only have a few big lists then this is decent work around.

**Comment by simkessy** [comment]
@dmccown1500  could you post a demo of what you did?

**Comment by dmccown1500** [comment]
Hey sorry did not have a ton of time, but I threw together a plunker. It is not quite working right but should at least give you some direction: http://plnkr.co/edit/OwhFCyHz0mO1yZMj5UW5?p=preview

**Comment by dmccown1500** [comment]
Update ok I got it working now: http://plnkr.co/edit/OwhFCyHz0mO1yZMj5UW5?p=preview I realized we were using a modified version of the library. pretty small change but it was slightly different. All credi to the ng-infinite-scroll team for the bulk of the code

**Comment by twrk** [comment]
@dmccown1500 thank you!! your plunker is very helpful :)

**Comment by dmccown1500** [comment]
@twrk glad to hear is was helpful. I love this library and would really like to see something like this get pulled in natively.

**Comment by dometec** [comment]
@dmccown1500 thank you too!! it's magic!

**Comment by igorlino** [comment]
@dmccown1500  +10  cool stuff. 

@twrk +1 making native this code

**Comment by igorlino** [comment]
@dmccown1500 I both the infinitescroll and the #430 are needed.

**Comment by Anton-Rodionov** [comment]
@jpdesigndev A bit about how to turn *ngAnimate*.
Every item of the list that ui-select renders is wrapped by ng-show. By default, every ng-show is animated (it might be not a real animation, but *ng-enter*, *ng-leave*, etc. classes still get added). That's what brings slowness. To avoid that you could define a regexp to animate only the elements you want.

```javascript
.config(['$animateProvider', function($animateProvider) {
        // Setup `ngAnimate` to work only with angular-motion classes (`am-` prefixed)
        // or custom animation classes (`animate-` prefixed) to boost performance
        $animateProvider.classNameFilter(/(animate|am)-/);
}])
```

**Comment by dmccown1500** [comment]
Anton, that is a really interesting idea. And it would be kind of nice in that it enforces the naming convention internally for animation classes. That said for library heavy apps it seems like it would cause a lot of animations in the libraries to not work. Thanks for the idea!

**Comment by jpdesigndev** [comment]
@Anton-Rodionov Thanks for the detailed explanation. This is exactly what I ended up doing. I'm still curious if there is a way to handle this for certain modules instead of app-wide. In any case, I got it working just as you showed. Thanks for leading me down the right track.

**Comment by hopewise** [comment]
how to turn ngAnimate off for all items of ui-select from inside the scope of where its been used?

**Comment by Anton-Rodionov** [comment]
@jpdesigndev @hopewise To restrict ngAnimate to the certain scope instead of whole application you could try `ng-animate-children="false"` directive.

**Comment by hopewise** [comment]
thanks, but it still very slow, I wonder how this custom-select 
https://github.com/axel-zarate/js-custom-select
 is very speedy comparing to select-ui

**Comment by simison** [comment]
@hopewise I switched to [Angular-Chosen](https://github.com/localytics/angular-chosen), a bit more established than js-custom-select.

**Comment by hopewise** [comment]
thanks for the link, I will check it right away..

**Comment by raphaelduartepinheiro** [comment]
@hopewise, I've checked this one, but unfortunately it depends on jQuery.

**Comment by estvmachine** [comment]
Vote for a native version of ui-select with **ng-infinite-scroll** like suggested @dmccown1500, thanks your plunker works

**Comment by hyzhak** [comment]
@estvmachine https://github.com/hyzhak/ui-select-infinity thy this extension for ui-select. If it won't work to you feel free add issue or PR.

**Comment by esion** [comment]
solving #221 could help here

**Comment by andrewboni** [comment]
@dmccown1500 thanks for that snippet- helpful. Any thoughts on why it breaks when using `theme="selectize"`?. See example plunkr here: http://plnkr.co/edit/cvOlgWaoBmvwoRrtY8XX?p=preview

**Comment by R4GN4R0KX** [comment]
@dmccown1500  thanks but the infinity extension disables the search

**Comment by brenovieira** [comment]
@dmccown1500, @simison, @amcdnl
Yesterday I took a look at `[infinite scroll](https://github.com/sroze/ngInfiniteScroll)`, and it "solves" my problem.

I have an array of at least 2.000 complex objects, so I was using `limitTo : 30`, but pre-selecting was broke. Now with `infinite scroll` seems to work fine, fast and pre-select works.

But here is my concern, now:
As I just added `jQuery` to use `infinite scroll`, I'm thinking what is the advantage of using `ui-select` over `[angular chosen](https://github.com/localytics/angular-chosen)`.

The feature I need is just a field to select multi options (multi select).

What do you guys think ?

**Comment by dmccown1500** [comment]
So our team used angular-chosen for a while and while it works 
reasonably well it seemed to have more issues with Angular proper and 
would occasionally not update correctly or have odd view problems when 
ng-show/ng-hide. If you did not have a jQuery dependency before I would 
think carefully if the infinite scroll solution is the right way to go, 
on the other hand almost inevitably you will eventually have a jQuery 
dependency on a large project so maybe its ok.

Breno Vieira wrote:

**Comment by dmccown1500** [comment]
@brenovieira So our team used angular-chosen for a while and while it works reasonably well it seemed to have more issues with Angular proper and would occasionally not update correctly or have odd view problems when ng-show/ng-hide. If you did not have a jQuery dependency before I would think carefully if the infinite scroll solution is the right way to go, on the other hand almost inevitably you will eventually have a jQuery dependency on a large project so maybe it is ok.

**Comment by brenovieira** [comment]
Thanks @dmccown1500 

I didn't know `angular chosen` was kind of "buggy".
When I searched for a lib to solve my multi select need some months ago, I thought `ui-select` was the right choise.

But now I don't know what to do, because `ui-select` has
- 5 or more open issues about performance, like this one or #648, #489, #221, #88, ...
- 8 or more open issues about `track by + multi select` or `dupes`, like #806 , #781, #366, #1223, #1233, #665, ...

Yesterday I looked all open issues and I think these two are the biggest issues with `ui-select` right now and my app has both hahaha

I really don't wanna add `jQuery`, since I'm building a chrome extension that just needs a [multi select datepicker](https://github.com/spongessuck/gm.datepickerMultiSelect) and a multi select field to select airports. Then the extension fetches flights for multi origins, destinations, departure and return dates. So its ui is isn't a large project with many dependencies.

What do you think @dimirc, @cdjackson, @ProLoser ?
Please, let me know if I can help with these two issues.

**Comment by ProLoser** [comment]
@brenovieira use a normal datepicker and just every time a selection is made, push to an array. I don't know why you're bothering with any other select widget.

**Comment by dmccown1500** [comment]
@brenovieira I would agree with @ProLoser this is way overkill for what you are trying to accomplish

**Comment by brenovieira** [comment]
@ProLoser, the datepicker need is already addressed.
Now I just need a field to select multi values that works fast with +2k items hahaha

Note: I just mentioned datepicker to show it's a "simple" application.
Note2: I can't use normal datepicker and $watch changes, because the datepicker doesn't show all selected dates.

I'm sorry for the confusion, but they're separated components on my app:

![extension](https://cloud.githubusercontent.com/assets/7158650/10649815/ce9d69a6-7824-11e5-8e18-1f658bfa8a06.png)

The datepicker works just fine after I made some PR.
But `ui-select` isn't ready for production in my case, because I have two ui-select. each has +2k complex items (all airports in the world) hehehhe

Did you get it now ?

**Comment by dmccown1500** [comment]
@brenovieira Ahh ok I see what you are trying to do now. So just thinking out loud here but Angular material has the idea of virtual scroll which might work with their chips, but it is tied to their library so that is probably a non-starter...

**Comment by brenovieira** [comment]
Yeah I saw it [here](https://material.angularjs.org/latest/api/directive/mdChips), but I didn't see an option to disable tagging.

I have a list of airports and the user must select one of them, not create a new.

**Comment by facesea** [comment]
i  solved same problem, when i annotation a line code of  `select.js`:
![qq20151217-1 2x](https://cloud.githubusercontent.com/assets/11919175/11868493/1d165188-a4f3-11e5-8eb4-c7afcb70b54c.png)

**Comment by ssljivic** [comment]
+1

**Comment by simkessy** [comment]
Did anyone ever solve this? I have a list of 700 items and I'm getting complaints about performance with no clue how to go about solving this.

**Comment by simkessy** [comment]
@dmccown1500 is there a way to use virtual repeat from ngmaterial with ui-select to fix the performance issues

**Comment by dmccown1500** [comment]
@simkessy I never tried I moved to a different project where the library was not in use so it kind of fell off my radar. Seems like it might work?

**Comment by AdamGerthel** [comment]
I just moved from oi-select to ui-select because oi-select doesn't support using scope variables to populate its attributes for some reason. Took me a good 3-4 hours to re-theme from oi to ui and rewrite some code.

And now this... gah!

**Comment by simkessy** [comment]
@AdamGerthel, this isn't a fix exactly, but I was able to help my performance by introducing a min-length and using the refresh directives built in. Essentially, I would only populate the dropdown when the user entered at least 2-3 characters. 

Pros: Significant performance improvement, instead of rendering 700 items I'm only showing a subset
Cons: You don't get a full list of values until you enter something

**Comment by AdamGerthel** [comment]
@simkessy thank you! It worked wonders. I followed this approach: https://github.com/angular-ui/ui-select/issues/88#issuecomment-179916133

**Comment by stringbeans** [comment]
@dmccown1500 thanks for providing your solution with infinite scroll! Very cool stuff :) Unfortunately it doesn't seem to be working with the latest version of ngInfiniteScroll as it attaches by default to the window and not the container. I tried using the exact copy of infinite-scroll you used in your plunkr demo and noticed that ui-select would redraw itself after increasing the limit dynamically, thus each time you scroll and it adds more items it skips back to the top of the list again. 

were you able to get your solution to work with the latest versions of infinite scroll and ui-select (ie. 0.14.x or 0.16.0) ?

**Comment by dmccown1500** [comment]
@stringbeans I left the project where I was using ui-select so I have not kept that up to date. So if you need a feature from one of the newer versions then you are kind of out of luck.

**Comment by TravisTam** [comment]
Is there any fix on this issue?

**Comment by user378230** [comment]
@TravisTam feel free to submit any PRs you feel could improve performance. 

I'll be happy to review :smiley:

**Comment by edriang** [comment]
Hi, wanted to share what I did in my project..

First commented the same line that mentioned @facesea:
`//            .attr('ng-if', '$select.open'); //Prevent unnecessary watches when dropdown is closed`

But also the major problem with ui-select is using it with `array | filter` syntax. As explained in this page [https://toddmotto.com/use-controller-filters-to-prevent-digest-performance-issues/](https://toddmotto.com/use-controller-filters-to-prevent-digest-performance-issues/) using filter in html causes big performance issues because it gets called on every digest cycle.

Then, if you don't use filter in view, `.attr('ng-if', '$select.open'); ` it's no longer necessary.

Finally I had to modify select.js to emit an "on-filter" event when `$select.search` changes, so I could update the source array when user types something.

```
$scope.$watch('$select.search', function(search){
        $scope.$emit('on-filter', search);
});
```

Hope this tips helps to improve the library.

**Comment by evenicoulddoit** [comment]
@stringbeans you're right, the latest infinite scroll plugin uses the window by default, but it *does* allow you to modify the container. The scope properties are:

```js
{
  infiniteScroll: '&',
  infiniteScrollContainer: '=',
  infiniteScrollDistance: '=',
  infiniteScrollDisabled: '=',
  infiniteScrollUseDocumentBottom: '=',
  infiniteScrollListenForEvent: '@'
}
```

However... I've tried in vain to set it up to work correctly. You need to pass a selector to `infiniteScrollContainer`, and neither using `.ui-select-choices-group` nor `#ui-select-choice-0` (which it looks like the old version was using automatically as the container) works.

**Comment by user378230** [comment]
@jocluz this issue is up to date with all info posted so far, what further news do you require?

Feel free to submit any PRs you feel could improve performance I'll be happy to review 😃

**Comment by benwigley** [comment]
I also have found a solution similar to some above, but in to my eyes was a bit more straight forward.

In the view

```html
<ui-select-choices repeat="school in filterSchools($select.search)">
  <div ng-bind-html="school.name | highlight: $select.search"></div>
</ui-select-choices>
``

In the controller. I have a list of 3,000+ schools, and I have no need for the user to scroll through them so I return an empty array when there is no search term. When there is a search term I limit the array to 30 items.

```coffee
$scope.filterSchools = (search) ->
    return [] if !search
    limit = 30
    index = 0
    searchRegex = new RegExp(search, 'gi') # gi = global and case insensitive
    return $scope.schools.filter (school) ->
      if school.name.match(searchRegex) && index < limit
        index += 1
        return true
```

If you want the dropdown to look nice (invisible) when there no items in the list then you might want to edit the css of `. ui-select-dropdown` and remove the border.

**Comment by jacek213** [comment]
@evenicoulddoit I'm also having a hard time integring ng-infinite-scroll. I've tried to add infinite scroll attributes to ui-select-choices like:

    <ui-select-choices repeat="user.id as user in mmCtrl.users | filter: $select.search"
                       refresh="mmCtrl.onNextPage($select)"
                       refresh-delay="300"

                       infinite-scroll="mmCtrl.onNextPage($select, $event, true)"
                       infinite-scroll-distance='0'
                       infinite-scroll-container="#ui-select-choices-0"
                       infinite-scroll-parent="true"
                       infinite-scroll-use-document-bottom="false"
                       >

No success so far - probably because the container to which I'm pointing is not really a parent of ui-select-choices but it's the element to be scrolled itself.

**Comment by kjsalexander** [comment]
After updating to the latest version of ui-select (0.19.x), I no longer have a performance issue with large arrays. Is this still an issue?

**Comment by Remiz** [comment]
@benwigley I tried something similar to you, but in my case I have 100k records in a list... The dropdown gets populated really fast and it seems to work. I'm using a multiple select and when I select more than one object it transform all the objects to the one selected last... I think the binding can't work in my case but if I bind to the full list the performances are terrible. Not sure what to do.

**Comment by jacek213** [comment]
@Remiz if you want to stick to ui-select, hooking up infinite scroll seems like the only option imho.

**Comment by 7899** [comment]
Infinite scroll solution is misleading.  the problem still persists, if you keep scrolling down the performance will start to degrade again as the list grows and grows.  

You can use LimitTo and Begin with dynamic variables that change when you hit ScrollTop and when you scroll to the bottom.  But now I have a problem where the search only operates on what is in the list (restricted by the LimitTo) and not on the complete array.  Is there a way to get it to search the underlying array instead?  Any advice appreciated, I'm very close to having a solution to this problem!

**Comment by sshalayev** [comment]
IMHO, root of the problem is in intensive and undjustify use of forced digesting. Also I'd say that ngModel is used rather strange way. I didn't find a single $commitViewValue in the code.
Making long things short. After removing unnecessary calls to digest and doing small tweaks on how ngModel and $compile are used, I've got it working with suitable performance on large arrays. It's not ready for PR perhaps - I've tested only functionality I need. Hopefully someone can get good insights out of it

**Comment by user378230** [comment]
@sshalayev it's a real shame you couldn't contribute your changes back to us here, I'm sure a lot of people would have been grateful if you did. 

Consider submitting them in smaller chunks if you want, it will save you having to constantly keep your own fork up to date after all! 😁

**Comment by cristianocoelho** [comment]
Adding comments here since I'm also interested in a way to speed up performance, 200 rows already is a killer.

**Comment by gercheq** [comment]
@cristianocoelho @user378230 Checkout this one for a working version: https://github.com/angular-ui/ui-select/issues/88#issuecomment-229049936

**Comment by gscoppino** [comment]
We've also been having some performance issues with `ui-select`, specifically with lists of over 100 items. Disabling AngularJS driven animations by removing the `ngAnimate` module significantly improves perceived performance as noted by @Anton-Rodionov earlier. However, we only want to disable animations for `ui-select` and want other parts of our app to continue using AngularJS driven animations. A solution was found here in this issue description: https://github.com/angular-ui/ui-select/issues/1502

In addition, there is a way to make the change on a per-instance basis and without forking the repository or keeping a locally modified copy of the library. We created a directive that requires `uiSelect` as a sibling and sets the instance of `$animate` to `null`, like so:

```
angular.module('my-module')
    .directive('uiSelectNoAnimate', () => ({
        restrict: 'A',
        require: 'uiSelect',
        link: (scope, element, attrs, $select) => {
            $select.$animate = null;
        }
    }));
```

and use as follows:
```
<ui-select ui-select-no-animate ...>...</ui-select>
```
